packages/router/src/adapters/contract/contract.ts

```diff
@@ -649,6 +762,15 @@ export const cancelRouterContract = async (
     },
     requestContext,
   );
+
+  incrementGasConsumed(
+    chainId,
+    receipt,
+    txData.sendingChainId === chainId ? TransactionReasons.CancelSender : TransactionReasons.CancelReceiver,
+    requestContext,
+  );
```

Comment:
Okay soooo this kinda sucks but, even with estimateGas calls beforehand, it's still possible for the sent transaction to revert onchain - which means we spent the gas (likely edge case I can think of is on a redundant call after a router reboot).

However, if the transaction reverts onchain, txService.sendTx would throw, meaning this block would not be reached.

So we need to have a try/catch/finally here, where in the finally block we incrementGasConsumed if either we got the receipt back (was successful) or the error thrown was from the transaction reverting onchain

Which is now making me realize that txservice should *probably* distinguish between an onchain revert (meaning gas  money was spent) and an offchain/gas estimate revert...

Reply:
For the last thing, we could just have TransactionReverted have an optional receipt member that gets set to the mined receipt in the event it reverts onchain

Reply:
Yes, but is there a way to check from the txservice the total gas consumed for a transaction (including bumping, retries, reverts etc)? I am aware of this edgecase, but not sure if there's currently a way to track it.

Polling the signer for native funds downed isn't really a great option, since there's no way to see which transactions are using the most gas

Reply:
Also, this problem exists in gelato as well. I'm not sure how the reverted transactions are handled there cc @rhlsthrm

Reply:
The problem exists for the gelato / relayer logic only with the above implementation. Just wait for the receipt, if we get it back, increment fees paid (bc then we *know* the fees were paid)

The only reason the edge case will persist here is bc we are explictly throwing if the tx was reverted on chain, but need to catch that and do an increment anyway in that event.

And yes, if the transaction is reverted onchain, the OnchainTransaction object internal to txservice grabs the (revert tx) receipt anyway, so we might as well use it.

Reply:
You're right -- totally thought that we should be paying them even in reverts (which we should) but that's not how its implemented. Let's just return it as part of the error then -- can work on that tomorrow (if @rhlsthrm doesn't get it done)

Reply:
We don't pay them in reverts, I think that's a risk that relayers take on themselves. They need to basically stop from getting frontrun.